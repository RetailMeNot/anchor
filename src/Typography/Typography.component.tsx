// REACT
import * as React from 'react';
// VENDOR
import classNames from 'classnames';
import styled, { css } from '@xstyled/styled-components';
import { th, variant } from '@xstyled/system';
// ANCHOR
import { space as spaceStyles, SpaceProps } from '@xstyled/system';
import { colors, Color } from '../theme';
import { TypographyTags, Scale } from '../theme/typography.theme';
import { rem } from '../utils/rem/rem';

// export type TypographyTags =
//     | 'a'
//     | 'p'
//     | 'span'
//     | 'h1'
//     | 'h2'
//     | 'h3'
//     | 'h4'
//     | 'h5'
//     | 'h6'
//     | 'blockquote'
//     | 'address'
//     | 'code'
//     | 'label'
//     | 'pre';

type FontWeights =
    | 'normal'
    | 'bold'
    | 'bolder'
    | 'lighter'
    | 100
    | 200
    | 300
    | 400
    | 500
    | 600
    | 700
    | 800
    | 900
    | 'initial'
    | 'inherit';

type TextTransforms =
    | 'none'
    | 'capitalize'
    | 'uppercase'
    | 'lowercase'
    | 'initial'
    | 'inherit';

type DisplayValues =
    | 'inline'
    | 'block'
    | 'contents'
    | 'flex'
    | 'grid'
    | 'inline-block'
    | 'inline-flex'
    | 'inline-grid'
    | 'inline-table'
    | 'list-item'
    | 'run-in'
    | 'table'
    | 'table-caption'
    | 'table-column-group'
    | 'table-header-group'
    | 'table-footer-group'
    | 'table-row-group'
    | 'table-cell'
    | 'table-column'
    | 'table-row'
    | 'none'
    | 'initial'
    | 'inherit';

export type TextAlign = 'center' | 'left' | 'right' | 'inherit' | 'justify';

// type Scale = 62 | 52 | 44 | 36 | 32 | 28 | 24 | 20 | 18 | 16 | 14 | 12;

export interface TypographyProps extends SpaceProps, React.HTMLAttributes<any> {
    className?: string;
    align?: TextAlign;
    transform?: TextTransforms;
    display?: DisplayValues;
    htmlFor?: string;
    href?: string;
    children?: any;
    tag?: TypographyTags;
    weight?: FontWeights;
    color?: 'inherit' | 'initial' | Color | string;
    hue?: 'light' | 'base' | 'dark';
    scale?: Scale;
    size?: number;
    lineHeight?: number;
}

const StyledTypography = (tag: TypographyTags) => styled(tag)<TypographyProps>`
    box-sizing: border-box;
    // Global Font Properties
    font-family: ${th('typography.fontFamily')};
    font-size: ${th('typography.fontSize')};
    font-weight: ${th('typography.fontWeight')};
    line-height: ${th('typography.lineHeight')};

    // Use a scale to set size & line-height
    ${variant({
        key: 'typography.scale',
        default: 16,
        prop: 'scale',
    })}

    // Apply default styles for element
    ${variant({
        key: 'typography.tag',
        default: 'span',
        prop: 'tag',
    })}

    // CSS Overrides
    ${({ align = 'inherit', display, transform = 'none' }) =>
        css({
            textAlign: align,
            display: display,
            textTransform: transform,
        })};

    // Override Size & Line Height
    ${({ lineHeight, size, weight }) =>
        css({
            lineHeight: rem(lineHeight),
            fontSize: rem(size),
            fontWeight: weight,
        })};

    // Spacing
    ${spaceStyles}

    // TODO: colors when theme colors are defined
    ${({ color = 'inherit', hue = 'base' }) =>
        css({ color: colors[color] ? colors[color][hue] : color })};

    small {
        font-size: 87.5%;
    }
`;

export const Typography = ({
    className,
    children,
    tag = 'span',
    ...props
}: TypographyProps): JSX.Element =>
    React.createElement(StyledTypography(tag), {
        children,
        tag,
        className: classNames('anchor-typography', className),
        ...props,
    });
