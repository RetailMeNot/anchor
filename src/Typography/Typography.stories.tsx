// REACT
import * as React from 'react';
// STORYBOOK
import { storiesOf } from '@storybook/react';
// VENDOR
import styled from 'styled-components';
import { ThemeProvider } from '@xstyled/styled-components';
// COMPONENTS
import { Typography } from './Typography.component';
import { Grid, Cell } from '../Grid';
// README
import * as README from './README.md';
// THEME
import typography from '../theme/typography.theme';

const StyledStory = styled.div`
    padding: 2rem 5rem;
`;

const scales = [62, 52, 44, 36, 32, 28, 24, 20, 18, 16, 14, 12];

const elements = [
    'p',
    'span',
    'h1',
    'h2',
    'h3',
    'h4',
    'h5',
    'h6',
    'blockquote',
    'address',
    'code',
    'pre',
];

const Theme = {
    typography,
};

storiesOf('Components/Typography', module)
    .addParameters({
        readme: {
            sidebar: README,
        },
    })
    .add('Default', () => (
        <ThemeProvider theme={Theme}>
            <StyledStory>
                <Typography tag="h1">Scales</Typography>
                <Grid columns="100%" minRowHeight="5rem" gap="1rem">
                    {scales.map((scale: any) => (
                        <Cell key={scale}>
                            <Typography color="ash" scale={scale}>
                                Font size is {scale}px | Scale value = {scale}
                            </Typography>
                        </Cell>
                    ))}
                </Grid>
                <Typography tag="h1">Elements</Typography>
                <Grid columns="100%" minRowHeight="5rem" gap="1rem">
                    {elements.map((element: any) => (
                        <Cell key={element}>
                            <Typography hue="dark" tag={element}>
                                Element: {element}
                            </Typography>
                        </Cell>
                    ))}
                </Grid>
            </StyledStory>
        </ThemeProvider>
    ));
