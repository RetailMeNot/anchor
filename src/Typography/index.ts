export { default as Typography } from './Typography.component';
