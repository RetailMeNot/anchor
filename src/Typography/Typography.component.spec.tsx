// REACT
import * as React from 'react';
import * as renderer from 'react-test-renderer';
// COMPONENT
import Typography from './Typography.component';
// ENZYME
import { mount } from 'enzyme';

// TEST SETUP
const subject = <Typography>Test text</Typography>;

describe('Component: Typography', () => {
    it('match its snapshot', () => {
        const tree = renderer.create(subject).toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('should render all text elements', () => {
        [
            'p',
            'span',
            'h1',
            'h2',
            'h3',
            'h4',
            'h5',
            'h6',
            'blockquote',
            'address',
            'code',
            'pre',
        ].forEach((element: any) => {
            expect(
                mount(<Typography as={element}>Test text</Typography>).exists(
                    element
                )
            ).toBeTruthy();
        });
    });
    it('should a span by default', () => {
        expect(
            mount(<Typography>Test text</Typography>).exists('span')
        ).toBeTruthy();
    });
});
