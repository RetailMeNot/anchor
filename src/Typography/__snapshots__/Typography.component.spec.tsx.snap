// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Typography match its snapshot 1`] = `
<span
  className="anchor-typography sc-bdVaJa lofSpu"
>
  Test text
</span>
`;
